File: update-zen-browser.py
#!/usr/bin/env python3

import os
import re
import sys
import requests
import subprocess
from datetime import datetime

# Configuration with environment variable support
def get_rpmbuild_path():
    """Get the RPM build path, supporting different environments"""
    # First check if RPM_BUILD_ROOT environment variable is set
    if "RPM_BUILD_ROOT" in os.environ:
        return os.environ["RPM_BUILD_ROOT"]

    # For GitHub Actions running in Fedora container
    if os.path.exists("/root/rpmbuild"):
        return "/root/rpmbuild"

    # Default to user's home directory
    return os.path.expanduser("~/rpmbuild")

# Set paths based on environment
RPMBUILD_PATH = get_rpmbuild_path()
GITHUB_API_URL = "https://api.github.com/repos/zen-browser/desktop/releases/latest"
SPEC_FILE_PATH = os.path.join(RPMBUILD_PATH, "SPECS/zen-browser.spec")
SOURCES_DIR = os.path.join(RPMBUILD_PATH, "SOURCES")
SRPMS_DIR = os.path.join(RPMBUILD_PATH, "SRPMS")

def get_latest_release():
    """Get the latest release info from GitHub API"""
    response = requests.get(GITHUB_API_URL)
    if response.status_code != 200:
        print(f"Error accessing GitHub API: {response.status_code}")
        sys.exit(1)

    release_data = response.json()
    version = release_data["tag_name"]

    # Skip twilight/nightly builds (containing 't' in version)
    if 't' in version:
        print(f"Skipping twilight/nightly build version: {version}")
        sys.exit(0)

    # Find the Linux x86_64 asset
    linux_asset = None
    for asset in release_data["assets"]:
        if "linux-x86_64.tar.xz" in asset["name"]:
            linux_asset = asset
            break

    if not linux_asset:
        print("Could not find Linux x86_64 asset in the release")
        sys.exit(1)

    return {
        "version": version,
        "download_url": f"https://github.com/zen-browser/desktop/releases/download/{version}/zen.linux-x86_64.tar.xz",
        "filename": "zen.linux-x86_64.tar.xz",
        "published_at": release_data["published_at"]
    }

def update_spec_file(release_info):
    """Update the spec file with the new version"""
    with open(SPEC_FILE_PATH, "r") as f:
        spec_content = f.read()

    # Update main version
    spec_content = re.sub(r"Version:\s+.*", f"Version:        {release_info['version']}", spec_content)

    # Update Source0 URL
    source_url = f"https://github.com/zen-browser/desktop/releases/download/{release_info['version']}/zen.linux-x86_64.tar.xz"
    spec_content = re.sub(r"Source0:\s+.*", f"Source0:        {source_url}", spec_content)

    # Update desktop entry version - Fixed version
    desktop_entry_pattern = r"\[Desktop Entry\]\nVersion=.*"
    spec_content = re.sub(desktop_entry_pattern,
                         f"[Desktop Entry]\nVersion={release_info['version']}",
                         spec_content,
                         flags=re.MULTILINE)

    # Add new changelog entry
    today = datetime.now().strftime("%a %b %d %Y")
    changelog_entry = f"%changelog\n* {today} COPR Build System <copr-build@fedoraproject.org> - {release_info['version']}-1\n- Update to {release_info['version']}\n"
    spec_content = re.sub(r"%changelog.*", changelog_entry, spec_content, flags=re.DOTALL)

    # Write the updated content back
    with open(SPEC_FILE_PATH, "w") as f:
        f.write(spec_content)

    return spec_content

def download_source(release_info):
    """Download the source tarball"""
    # Ensure the SOURCES directory exists
    os.makedirs(SOURCES_DIR, exist_ok=True)

    source_path = os.path.join(SOURCES_DIR, release_info["filename"])
    response = requests.get(release_info["download_url"], stream=True)

    if response.status_code != 200:
        print(f"Error downloading source: {response.status_code}")
        sys.exit(1)

    with open(source_path, "wb") as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)

    return source_path

def build_srpm():
    """Build the SRPM package"""
    result = subprocess.run(
        ["rpmbuild", "-bs", SPEC_FILE_PATH],
        capture_output=True,
        text=True
    )

    if result.returncode != 0:
        print(f"Error building SRPM: {result.stderr}")
        sys.exit(1)

    srpm_path = find_srpm_in_output(result)
    if not srpm_path:
        srpm_path = find_srpm_in_spec()
    if not srpm_path:
        srpm_path = find_srpm_in_directory()

    if not srpm_path:
        print("Could not find built SRPM path in output")
        print("stdout:", result.stdout)
        print("stderr:", result.stderr)
        sys.exit(1)

    print(f"Found SRPM: {srpm_path}")
    return srpm_path

def find_srpm_in_output(result):
    """Extract SRPM path from command output"""
    for line in result.stderr.split("\n"):
        if line.endswith(".src.rpm"):
            return line.strip().replace("Wrote: ", "")

    for line in result.stdout.split("\n"):
        if line.endswith(".src.rpm"):
            return line.strip().replace("Wrote: ", "")

    return None

def find_srpm_in_spec():
    """Find SRPM based on spec file version info"""
    with open(SPEC_FILE_PATH, "r") as f:
        spec_content = f.read()
        version_match = re.search(r"Version:\s+(.*)", spec_content)
        release_match = re.search(r"Release:\s+(.*)", spec_content)

        if version_match and release_match:
            version = version_match.group(1)
            release = release_match.group(1).replace("%{?dist}", ".fc41")
            expected_path = os.path.join(SRPMS_DIR, f"zen-browser-{version}-{release}.src.rpm")
            if os.path.exists(expected_path):
                return expected_path
    return None

def find_srpm_in_directory():
    """Find most recent SRPM in SRPMS directory"""
    try:
        os.makedirs(SRPMS_DIR, exist_ok=True)
        for file in os.listdir(SRPMS_DIR):
            if file.endswith(".src.rpm"):
                print(f" - {file}")
                return os.path.join(SRPMS_DIR, file)
    except Exception as e:
        print(f"Error listing SRPMS directory: {e}")
    return None

def submit_to_copr(srpm_path):
    """Submit the SRPM to COPR for building"""
    copr_project = "51ddh4r7h/zen-browser"

    # Strip "Wrote: " prefix if present
    if srpm_path.startswith("Wrote: "):
        srpm_path = srpm_path.replace("Wrote: ", "")

    print(f"Submitting {srpm_path} to COPR project {copr_project}...")

    result = subprocess.run(
        ["copr-cli", "build", copr_project, srpm_path],
        capture_output=True,
        text=True
    )

    if result.returncode != 0:
        print(f"Error submitting to COPR: {result.stderr}")
        sys.exit(1)

    print(f"Successfully submitted to COPR: {result.stdout}")

    # Extract the build ID from the output
    build_id_match = re.search(r"Created builds: (\d+)", result.stdout)
    if build_id_match:
        build_id = build_id_match.group(1)
        print(f"Build ID: {build_id}")
        print(f"Build status URL: https://copr.fedorainfracloud.org/coprs/build/{build_id}/")

def main():
    print("Checking for new Zen Browser releases...")
    latest_release = get_latest_release()

    # Check if this is a new version
    with open(SPEC_FILE_PATH, "r") as f:
        spec_content = f.read()
        version_match = re.search(r"Version:\s+(.*)", spec_content)
        if not version_match:
            print("Error: Could not find Version in spec file")
            sys.exit(1)
        current_version = version_match.group(1)

    if current_version == latest_release["version"]:
        print(f"Already at the latest version: {current_version}")
        return

    print(f"New version found: {latest_release['version']}")
    print("Downloading source...")
    download_source(latest_release)

    print("Updating spec file...")
    update_spec_file(latest_release)

    print("Building SRPM...")
    srpm_path = build_srpm()

    print("Submitting to COPR...")
    submit_to_copr(srpm_path)

    print("Done!")

if __name__ == "__main__":
    main()

File: update-zen-browser.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// Configuration and constant definitions
const (
	githubAPIURL = "https://api.github.com/repos/zen-browser/desktop/releases/latest"
	coprProject  = "51ddh4r7h/zen-browser"
)

// ReleaseInfo stores the release information from GitHub
type ReleaseInfo struct {
	Version     string
	DownloadURL string
	Filename    string
	PublishedAt string
}

// GitHubRelease represents the GitHub release API response structure
type GitHubRelease struct {
	TagName     string  `json:"tag_name"`
	PublishedAt string  `json:"published_at"`
	Assets      []Asset `json:"assets"`
}

// Asset represents a release asset from GitHub
type Asset struct {
	Name        string `json:"name"`
	DownloadURL string `json:"browser_download_url"`
}

// Get the RPM build path, supporting different environments
func getRpmbuildPath() string {
	// First check if RPM_BUILD_ROOT environment variable is set
	if rpmBuildRoot, exists := os.LookupEnv("RPM_BUILD_ROOT"); exists {
		return rpmBuildRoot
	}

	// For GitHub Actions running in Fedora container
	if _, err := os.Stat("/root/rpmbuild"); err == nil {
		return "/root/rpmbuild"
	}

	// Default to user's home directory
	homeDir, err := os.UserHomeDir()
	if err != nil {
		fmt.Println("Error getting home directory:", err)
		os.Exit(1)
	}
	return filepath.Join(homeDir, "rpmbuild")
}

// GetLatestRelease fetches the latest release information from GitHub
func getLatestRelease() (*ReleaseInfo, error) {
	resp, err := http.Get(githubAPIURL)
	if err != nil {
		return nil, fmt.Errorf("error accessing GitHub API: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("error accessing GitHub API: %d", resp.StatusCode)
	}

	var release GitHubRelease
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return nil, fmt.Errorf("error parsing GitHub API response: %v", err)
	}

	version := release.TagName

	// Skip twilight/nightly builds (containing 't' in version)
	if strings.Contains(version, "t") {
		fmt.Printf("Skipping twilight/nightly build version: %s\n", version)
		return nil, nil
	}

	// Find the Linux x86_64 asset
	var linuxAssetURL string
	for _, asset := range release.Assets {
		if strings.Contains(asset.Name, "linux-x86_64.tar.xz") {
			linuxAssetURL = asset.DownloadURL
			break
		}
	}

	if linuxAssetURL == "" {
		return nil, fmt.Errorf("could not find Linux x86_64 asset in the release")
	}

	return &ReleaseInfo{
		Version:     version,
		DownloadURL: fmt.Sprintf("https://github.com/zen-browser/desktop/releases/download/%s/zen.linux-x86_64.tar.xz", version),
		Filename:    "zen.linux-x86_64.tar.xz",
		PublishedAt: release.PublishedAt,
	}, nil
}

// UpdateSpecFile updates the spec file with the new version information
func updateSpecFile(specFilePath string, releaseInfo *ReleaseInfo) error {
	content, err := os.ReadFile(specFilePath)
	if err != nil {
		return fmt.Errorf("error reading spec file: %v", err)
	}

	// Update main version
	versionRegex := regexp.MustCompile(`Version:\s+.*`)
	updatedContent := versionRegex.ReplaceAllString(string(content), fmt.Sprintf("Version:        %s", releaseInfo.Version))

	// Update Source0 URL
	sourceURL := fmt.Sprintf("https://github.com/zen-browser/desktop/releases/download/%s/zen.linux-x86_64.tar.xz", releaseInfo.Version)
	sourceRegex := regexp.MustCompile(`Source0:\s+.*`)
	updatedContent = sourceRegex.ReplaceAllString(updatedContent, fmt.Sprintf("Source0:        %s", sourceURL))

	// Update desktop entry version
	desktopEntryRegex := regexp.MustCompile(`\[Desktop Entry\]\nVersion=.*`)
	updatedContent = desktopEntryRegex.ReplaceAllString(updatedContent, fmt.Sprintf("[Desktop Entry]\nVersion=%s", releaseInfo.Version))

	// Add new changelog entry
	today := time.Now().Format("Mon Jan 2 2006")
	changelogEntry := fmt.Sprintf("%%changelog\n* %s COPR Build System <copr-build@fedoraproject.org> - %s-1\n- Update to %s\n",
		today, releaseInfo.Version, releaseInfo.Version)
	changelogRegex := regexp.MustCompile(`%changelog.*`)
	updatedContent = changelogRegex.ReplaceAllString(updatedContent, changelogEntry)

	// Write the updated content back
	return os.WriteFile(specFilePath, []byte(updatedContent), 0644)
}

// DownloadSource downloads the source tarball
func downloadSource(sourcesDir, downloadURL, filename string) (string, error) {
	// Ensure the SOURCES directory exists
	if err := os.MkdirAll(sourcesDir, 0755); err != nil {
		return "", fmt.Errorf("error creating SOURCES directory: %v", err)
	}

	sourcePath := filepath.Join(sourcesDir, filename)

	// Download the file
	resp, err := http.Get(downloadURL)
	if err != nil {
		return "", fmt.Errorf("error downloading source: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("error downloading source: %d", resp.StatusCode)
	}

	file, err := os.Create(sourcePath)
	if err != nil {
		return "", fmt.Errorf("error creating source file: %v", err)
	}
	defer file.Close()

	_, err = io.Copy(file, resp.Body)
	if err != nil {
		return "", fmt.Errorf("error saving source file: %v", err)
	}

	return sourcePath, nil
}

// BuildSRPM builds the SRPM package
func buildSRPM(specFilePath string) (string, error) {
	cmd := exec.Command("rpmbuild", "-bs", specFilePath)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("error building SRPM: %v\nStderr: %s", err, stderr.String())
	}

	// Try to find the SRPM path from the output
	srpmPath := findSRPMInOutput(stdout.String(), stderr.String())
	if srpmPath == "" {
		srpmPath = findSRPMInSpec(specFilePath)
	}
	if srpmPath == "" {
		srpmPath = findSRPMInDirectory(filepath.Join(filepath.Dir(filepath.Dir(specFilePath)), "SRPMS"))
	}

	if srpmPath == "" {
		return "", fmt.Errorf("could not find built SRPM path in output\nStdout: %s\nStderr: %s",
			stdout.String(), stderr.String())
	}

	fmt.Printf("Found SRPM: %s\n", srpmPath)
	return srpmPath, nil
}

// FindSRPMInOutput extracts SRPM path from command output
func findSRPMInOutput(stdout, stderr string) string {
	// First check stderr
	scanner := bufio.NewScanner(strings.NewReader(stderr))
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasSuffix(line, ".src.rpm") {
			return strings.TrimPrefix(strings.TrimSpace(line), "Wrote: ")
		}
	}

	// Then check stdout
	scanner = bufio.NewScanner(strings.NewReader(stdout))
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasSuffix(line, ".src.rpm") {
			return strings.TrimPrefix(strings.TrimSpace(line), "Wrote: ")
		}
	}

	return ""
}

// FindSRPMInSpec finds SRPM based on spec file version info
func findSRPMInSpec(specFilePath string) string {
	content, err := os.ReadFile(specFilePath)
	if err != nil {
		return ""
	}

	// Extract version
	versionRegex := regexp.MustCompile(`Version:\s+(.*)`)
	versionMatches := versionRegex.FindStringSubmatch(string(content))

	// Extract release
	releaseRegex := regexp.MustCompile(`Release:\s+(.*)`)
	releaseMatches := releaseRegex.FindStringSubmatch(string(content))

	if len(versionMatches) > 1 && len(releaseMatches) > 1 {
		version := versionMatches[1]
		release := strings.Replace(releaseMatches[1], "%{?dist}", ".fc41", 1)

		srpmDir := filepath.Join(filepath.Dir(filepath.Dir(specFilePath)), "SRPMS")
		expectedPath := filepath.Join(srpmDir, fmt.Sprintf("zen-browser-%s-%s.src.rpm", version, release))

		if _, err := os.Stat(expectedPath); err == nil {
			return expectedPath
		}
	}

	return ""
}

// FindSRPMInDirectory finds most recent SRPM in SRPMS directory
func findSRPMInDirectory(srpmsDir string) string {
	if err := os.MkdirAll(srpmsDir, 0755); err != nil {
		fmt.Printf("Error creating SRPMS directory: %v\n", err)
		return ""
	}

	files, err := os.ReadDir(srpmsDir)
	if err != nil {
		fmt.Printf("Error listing SRPMS directory: %v\n", err)
		return ""
	}

	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".src.rpm") {
			fmt.Printf(" - %s\n", file.Name())
			return filepath.Join(srpmsDir, file.Name())
		}
	}

	return ""
}

// SubmitToCopr submits the SRPM to COPR for building
func submitToCopr(srpmPath string) error {
	// Strip "Wrote: " prefix if present
	srpmPath = strings.TrimPrefix(srpmPath, "Wrote: ")

	fmt.Printf("Submitting %s to COPR project %s...\n", srpmPath, coprProject)

	cmd := exec.Command("copr-cli", "build", coprProject, srpmPath)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("error submitting to COPR: %v\nStderr: %s", err, stderr.String())
	}

	fmt.Printf("Successfully submitted to COPR: %s\n", stdout.String())

	// Extract the build ID from the output
	buildIDRegex := regexp.MustCompile(`Created builds: (\d+)`)
	buildIDMatches := buildIDRegex.FindStringSubmatch(stdout.String())

	if len(buildIDMatches) > 1 {
		buildID := buildIDMatches[1]
		fmt.Printf("Build ID: %s\n", buildID)
		fmt.Printf("Build status URL: https://copr.fedorainfracloud.org/coprs/build/%s/\n", buildID)
	}

	return nil
}

func main() {
	fmt.Println("Checking for new Zen Browser releases...")

	// Set paths based on environment
	rpmbuildPath := getRpmbuildPath()
	specFilePath := filepath.Join(rpmbuildPath, "SPECS/zen-browser.spec")
	sourcesDir := filepath.Join(rpmbuildPath, "SOURCES")

	// Get latest release info
	releaseInfo, err := getLatestRelease()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	// Skip if we got nil due to twilight/nightly build
	if releaseInfo == nil {
		os.Exit(0)
	}

	// Check if this is a new version
	specContent, err := os.ReadFile(specFilePath)
	if err != nil {
		fmt.Printf("Error reading spec file: %v\n", err)
		os.Exit(1)
	}

	versionRegex := regexp.MustCompile(`Version:\s+(.*)`)
	versionMatches := versionRegex.FindStringSubmatch(string(specContent))

	if len(versionMatches) < 2 {
		fmt.Println("Error: Could not find Version in spec file")
		os.Exit(1)
	}

	currentVersion := versionMatches[1]

	if currentVersion == releaseInfo.Version {
		fmt.Printf("Already at the latest version: %s\n", currentVersion)
		return
	}

	fmt.Printf("New version found: %s\n", releaseInfo.Version)

	fmt.Println("Downloading source...")
	_, err = downloadSource(sourcesDir, releaseInfo.DownloadURL, releaseInfo.Filename)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	fmt.Println("Updating spec file...")
	err = updateSpecFile(specFilePath, releaseInfo)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	fmt.Println("Building SRPM...")
	srpmPath, err := buildSRPM(specFilePath)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	fmt.Println("Submitting to COPR...")
	err = submitToCopr(srpmPath)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	fmt.Println("Done!")
}
